### DISCUSSION TITLE
What is the recommended way to make a deep copy of a request in KP?

### DISCUSSION THREAD

**Discussion by joffinjoy** [Sunbird-inQuiry/Community#6] [Q&A] [answered]
This is a request to UpdateHierarchyManager.updateHierarchy(hierarchyRequest).

hierarchyRequest:

> Request[context = {
	graph_id = domain,
	schemaName = questionset,
	version = 1.0,
	objectType = QuestionSet
}, requestValueObjects = {
	nodesModified = {
		596 d944a - f2d3 - 4 a12 - 96 cb - 1e95796 d2820 = {
			setDefaultValue = false,
			metadata = {
				instructions = null,
				code = Q1,
				channel = {
					{
						channel_id
					}
				},
				description = Q1,
				responseDeclaration = null,
				language = [English],
				mimeType = application / vnd.sunbird.question,
				media = null,
				body = This is Body 1,
				editorState = null,
				objectType = Question,
				interactions = null,
				primaryCategory = Slider,
				contentDisposition = inline,
				contentEncoding = gzip,
				showSolutions = No,
				allowAnonymousAccess = Yes,
				lastStatusChangedOn = 2022 - 03 - 29 T15: 37: 42.837 + 0530,
				visibility = Parent,
				showTimer = No,
				solutions = null,
				hints = null,
				languageCode = [en],
				version = 1,
				showFeedback = No,
				license = CC BY 4.0,
				answer = This is Answer 1,
				name = Q1
			},
			root = false,
			isNew = true,
			objectType = Question
		},
		ab33aa8e - 04 f3 - 4 aa0 - a23b - faa37f5724fc = {
			setDefaultValue = false,
			metadata = {
				code = S1,
				allowSkip = Yes,
				containsUserData = No,
				channel = {
					{
						channel_id
					}
				},
				description = Section 1,
				language = [English],
				mimeType = application / vnd.sunbird.questionset,
				showHints = No,
				objectType = QuestionSet,
				primaryCategory = Observation,
				contentDisposition = inline,
				contentEncoding = gzip,
				generateDIALCodes = No,
				showSolutions = No,
				allowAnonymousAccess = Yes,
				lastStatusChangedOn = 2022 - 03 - 29 T15: 37: 42.872 + 0530,
				requiresSubmit = No,
				visibility = Parent,
				showTimer = No,
				setType = materialised,
				languageCode = [en],
				version = 1,
				showFeedback = No,
				license = CC BY 4.0,
				name = S1,
				navigationMode = non - linear,
				allowBranching = Yes,
				shuffle = true
			},
			root = false,
			isNew = true,
			objectType = QuestionSet
		}
	},
	hierarchy = {
		do_11350687937866137611 = {
			children = [ab33aa8e - 04 f3 - 4 aa0 - a23b - faa37f5724fc],
			primaryCategory = Observation,
			root = true
		},
		ab33aa8e - 04 f3 - 4 aa0 - a23b - faa37f5724fc = {
			children = [596 d944a - f2d3 - 4 a12 - 96 cb - 1e95796 d2820, do_1135049090076835841173],
			primaryCategory = Observation,
			root = false
		}
	}
}]
_Ignore any inconsistencies in the request itself. It's poorly parsed for presentation._

I want to a make a deep copy of this request before updateHierarchy() call is made. It has to be a deep copy because updateHierarchy() modifies the request during its processing and I no longer can use the original request or a shallow copy once updateHierarchy has finished its execution. What is the recommended way to do this?

Note: The goal is to use the same request with all nodes in nodesModified set as isNew=false, and replace all identifiers with appropriate do_ids and re-run the updateHierarchy().
@pallakartheekreddy @krgauraw

**Comment by krgauraw** [comment, marked as answer]
@joffinjoy In order to make a deep copy of Sunbird Request, you can serialize it and then deserialize back as a Request class object. For serialization and deserialization, you can use exiting files like JsonUtils or ScalaJsonUtils.

cc: @pallakartheekreddy

**Reply by Krishnaj20** [reply]
@krgauraw - thanks for responding to this question. Can you add this question to the component specific FAQ section in the inQuiry microsite as well? Thanks

**Reply by joffinjoy** [reply]
@krgauraw Thanks Gauraw... It worked. 

Posting for posterity:
```
import org.sunbird.common.{JsonUtils, Platform}
val newUpdateRequest = JsonUtils.deserialize(ScalaJsonUtils.serialize(hierarchyRequest), classOf[Request])
```

**Comment by pramodkvarma** [comment]
@Krishnaj20 you can update FAQ yourself right? I suggest you also do these doc/faq fixes (non code/tech portions) etc and become a contributor yourself (apologies if you are already doing non tech doc updates yourself).

**Reply by Krishnaj20** [reply]
Created a new [FAQ](https://app.gitbook.com/o/-Mi9QwJlsfb7xuxTBc0J/s/Wu4HIWGkb7dD4y0Kup4W/learn/product-and-developer-guide/question-and-question-set-service/faqs) section for question/question set services and added this question.

**Reply by pramodkvarma** [reply]
@Krishnaj20 Can you please share public link? This link needs permission. Thanks.

**Reply by Krishnaj20** [reply]
Apologies - posted the gitbook link. [Here](https://inquiry.sunbird.org/learn/product-and-developer-guide/question-and-question-set-service/faqs) is the public link.

**Reply by pramodkvarma** [reply]
Thanks @Krishnaj20. Good start. But I suggest two ideas:
1. Instead of just one advanced FAQ item, please add others and make it reasonably useful.
2. Instead of having entire long answer, FAQ answer should be just 2-3 sentences with a URL to details. In this case answer within FAQ can point to the discussion URL where it is answered.

**Reply by Krishnaj20** [reply]
thanks @pramodkvarma. I've edited the FAQ answer to be brief with a link to this discussion thread for additional details. I'll work with @krgauraw & @pallakartheekreddy to expand the FAQs.
